src/modules/content-s3/s3.c

```diff
@@ -0,0 +1,259 @@
+/************************************************************\
+ * Copyright 2020 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <sys/types.h>
+#include <string.h>
+#include <errno.h>
+#include <stdio.h>
+#include <libs3.h>
+#include <stdlib.h>
+
+#include "src/common/libutil/errno_safe.h"
+
+#include "s3.h"
+
+static S3Status status_g;
+static S3Protocol protocol = S3ProtocolHTTP;
+static S3UriStyle uri_style = S3UriStylePath;
+
+/* Data needed by the get object callback function
+ */
+struct get_cb_data {
+    size_t size;
+    void *data;
+};
+
+struct put_cb_data {
+    const void *data;
+    int count;
+};
+
+static S3Status response_props_cb(const S3ResponseProperties *properties, void *data)
+{
+    return S3StatusOK;
+}
+
+static void response_complete_cb(S3Status status, const S3ErrorDetails *error, void *data)
+{
+    status_g = status;
+}
+
+/* Writes a chunk of 'data' to s3, returning
+ * the size of the data written.
+ */
+static int put_object_cb(int buff_size, char *buff, void *data)
+{
+    struct put_cb_data *cb_data = data;
+
+    memcpy(buff, cb_data->data + cb_data->count, buff_size);
```

Comment:
To be defensive, we might want to include the put size in the `put_cb_data` struct and then make sure that the memcpy size is the lesser of `buff_size` and `(cb_data->size - cb_data->count)`.  And also return that size not `buff_size`.